crates/control/src/services/discovery.rs

```diff
@@ -0,0 +1,190 @@
+use std::collections::BTreeMap;
+
+use models::Capture;
+use models::CaptureBinding;
+use models::CaptureDef;
+use models::CaptureEndpoint;
+use models::Catalog;
+use models::Collection;
+use models::CollectionDef;
+use models::CompositeKey;
+use models::Config;
+use models::ConnectorConfig;
+use models::JsonPointer;
+use models::Object;
+use models::RelativeUrl;
+use models::Schema;
+use models::ShardTemplate;
+
+use crate::models::connector_images::ConnectorImage;
+use crate::models::connectors::Connector;
+use crate::services::connectors::DiscoveryOptions;
+
+#[derive(Debug, Deserialize)]
+pub struct DiscoverResponse {
+    pub bindings: Vec<DiscoveredBinding>,
+}
+
+#[derive(Debug, Deserialize)]
+#[serde(rename_all = "camelCase")]
+pub struct DiscoveredBinding {
+    /// A recommended display name for this discovered binding.
+    pub recommended_name: String,
+    /// JSON-encoded object which specifies the endpoint resource to be captured.
+    #[serde(rename = "resourceSpec")]
+    pub resource_spec_json: Object,
+    /// JSON schema of documents produced by this binding.
+    #[serde(rename = "documentSchema")]
+    pub document_schema_json: Object,
+    /// Composite key of documents (if known), as JSON-Pointers.
+    pub key_ptrs: Vec<String>,
+}
+
+impl DiscoveredBinding {
+    fn key(&self) -> models::CompositeKey {
+        CompositeKey::new(
+            self.key_ptrs
+                .iter()
+                .map(JsonPointer::new)
+                .collect::<Vec<JsonPointer>>(),
+        )
+    }
+
+    pub fn schema_url(&self) -> Schema {
+        Schema::Url(RelativeUrl::new(self.schema_name()))
+    }
+
+    pub fn schema_name(&self) -> String {
+        format!("{}.schema.json", self.recommended_name)
+    }
+}
+
+pub struct DiscoveredCatalog {
+    connector: Connector,
+    image: ConnectorImage,
+    config: Object,
+    bindings: Vec<DiscoveredBinding>,
```

Comment:
nit: Do we ever ensure that the `recommended_name` of each binding is unique? If not, then we probably should, and this might be a reasonable place to do that. Either that, or have `api discover` do it. I think just adding a `_1` (`_2`, etc) to the ends of duplicate names should do it. But probably best not to assume the connectors will all ensure uniqueness.